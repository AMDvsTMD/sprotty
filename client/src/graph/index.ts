export * from "./model/sgraph"
export * from "./model/sgraph-factory"
export * from "./view/views"
