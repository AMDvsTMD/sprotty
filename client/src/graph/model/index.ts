export * from './sgraph';
export * from './sgraph-factory';
