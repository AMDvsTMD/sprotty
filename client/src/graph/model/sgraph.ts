
import { SModelRootSchema, SParentElementSchema, SChildElement, SModelElementSchema } from "../../base/model/smodel"
import { Point, IDENTITY_MATRIX, Bounds, EMPTY_BOUNDS } from "../../utils/geometry"
import { ViewportRootElement } from "../../features/viewport/viewport-root"
import { selectFeature, Selectable } from "../../features/select/model"
import { moveFeature } from "../../features/move/model"
import { boundsFeature, BoundsAware } from "../../features/bounds/model"

export interface SGraphSchema extends SModelRootSchema {
    children: SGraphElementSchema[]
    width?: number
    height?: number
    scroll: Point
    zoom: number
}

export class SGraph extends ViewportRootElement implements SGraphSchema {
}

export interface SNodeSchema extends SParentElementSchema {
    x?: number
    y?: number
    width?: number
    height?: number
    children?: SGraphElementSchema[]
}

export class SNode extends SChildElement implements SNodeSchema, Selectable, BoundsAware {
    x: number = 0
    y: number = 0
    width: number = -1
    height: number = -1
    children: SGraphElement[]
    selected: boolean = false

    get bounds(): Bounds {
        return {x: this.x, y: this.y, width: this.width, height: this.height}
    }

    set bounds(bounds: Bounds) {
        this.x = bounds.x
        this.y = bounds.y
        this.width = bounds.width
        this.height = bounds.height
    }

    get position(): Point {
        return { x: this.x, y: this.y }
    }

    set position(point: Point) {
        this.x = point.x
        this.y = point.y
    }

    hasFeature(feature: symbol): boolean {
        return feature === selectFeature || feature === moveFeature || feature === boundsFeature
    }
}

export interface SEdgeSchema extends SModelElementSchema {
    sourceId: string
    targetId: string
    routingPoints?: Point[]
}

export class SEdge extends SChildElement implements SEdgeSchema {
    sourceId: string
    targetId: string
    routingPoints: Point[] = []

    get source(): SNode | undefined {
        return this.index.getById(this.sourceId) as SNode
    }

    get target(): SNode | undefined {
        return this.index.getById(this.targetId) as SNode
    }
}

export type SGraphElementSchema = SNodeSchema | SEdgeSchema
export type SGraphElement = SNode | SEdge
