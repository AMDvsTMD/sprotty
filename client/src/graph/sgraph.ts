/*
 * Copyright (C) 2017 TypeFox and others.
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
 */

import { SChildElement, SModelElementSchema, SModelRootSchema, SModelIndex, SModelElement } from '../base/model/smodel'
import {
    boundsFeature, layoutContainerFeature, layoutableChildFeature, Alignable, alignFeature, ModelLayoutOptions
} from '../features/bounds/model'
import { Fadeable, fadeFeature } from '../features/fade/model'
import { Hoverable, hoverFeedbackFeature, popupFeature } from '../features/hover/model'
import { moveFeature } from '../features/move/model'
import { Selectable, selectFeature } from '../features/select/model'
import { ViewportRootElement } from '../features/viewport/viewport-root'
import { Bounds, ORIGIN_POINT, Point } from '../utils/geometry'
import { SShapeElement, SShapeElementSchema } from '../features/bounds/model'
import { editFeature, Routable } from '../features/edit/model'

/**
 * Serializable schema for graph-like models.
 */
export interface SGraphSchema extends SModelRootSchema {
    children: SModelElementSchema[]
    bounds?: Bounds
    scroll?: Point
    zoom?: number
    layoutOptions?: ModelLayoutOptions
}

/**
 * Root element for graph-like models.
 */
export class SGraph extends ViewportRootElement {
    layoutOptions?: ModelLayoutOptions

    constructor(index = new SGraphIndex()) {
        super(index)
    }
}

/**
 * Serializable schema for SNode.
 */
export interface SNodeSchema extends SShapeElementSchema {
    layout?: string
    selected?: boolean
    hoverFeedback?: boolean
    opacity?: number
}

/**
 * Model element class for nodes, which are connectable entities in a graph. A node can be connected to
 * another node via an SEdge. Such a connection can be direct, i.e. the node is the source or target of
 * the edge, or indirect through a port, i.e. it contains an SPort which is the source or target of the edge.
 */
export class SNode extends SShapeElement implements Selectable, Fadeable, Hoverable {
    children: SChildElement[]
    layout?: string
    selected: boolean = false
    hoverFeedback: boolean = false
    opacity: number = 1

    hasFeature(feature: symbol): boolean {
        return feature === selectFeature || feature === moveFeature || feature === boundsFeature
            || feature === layoutContainerFeature || feature === fadeFeature || feature === hoverFeedbackFeature
            || feature === popupFeature
    }

    get incomingEdges(): Iterable<SEdge> {
        return (this.index as SGraphIndex).getIncomingEdges(this)
    }

    get outgoingEdges(): Iterable<SEdge> {
        return (this.index as SGraphIndex).getOutgoingEdges(this)
    }
}

/**
 * Serializable schema for SPort.
 */
export interface SPortSchema extends SShapeElementSchema {
    selected?: boolean
    opacity?: number
}

/**
 * A port is a connection point for edges. It should always be contained in an SNode.
 */
export class SPort extends SShapeElement implements Selectable, Fadeable, Hoverable {
    hoverFeedback: boolean = false
    selected: boolean = false
    opacity: number = 1

    hasFeature(feature: symbol): boolean {
        return feature === selectFeature || feature === boundsFeature || feature === fadeFeature
            || feature === hoverFeedbackFeature
    }
}

/**
 * Serializable schema for SEdge.
 */
export interface SEdgeSchema extends SModelElementSchema {
    sourceId: string
    targetId: string
    routingPoints?: Point[]
    opacity?: number
}
export interface SEdgeAnchorsSchema {
    sourceAnchor: Point
    targetAnchor: Point
}

/**
 * Model element class for edges, which are the connectors in a graph. An edge has a source and a target,
 * each of which can be either a node or a port. The source and target elements are referenced via their
 * ids and can be resolved with the index stored in the root element.
 */
export class SEdge extends SChildElement implements Fadeable, Selectable, Routable, Hoverable {
    hoverFeedback: boolean = false
    sourceId: string
    targetId: string
    routingPoints: Point[] = []
    opacity: number = 1
    selected: boolean = false

    get source(): SNode | SPort | undefined {
        return this.index.getById(this.sourceId) as SNode | SPort
    }

    get target(): SNode | SPort | undefined {
        return this.index.getById(this.targetId) as SNode | SPort
    }

    hasFeature(feature: symbol): boolean {
        return feature === fadeFeature || feature === selectFeature ||
            feature === editFeature || feature === hoverFeedbackFeature
    }
}

/**
 * Serializable schema for SLabel.
 */
export interface SLabelSchema extends SShapeElementSchema {
    text: string
    selected?: boolean
}

/**
 * A label can be attached to a node, edge, or port, and contains some text to be rendered in its view.
 */
export class SLabel extends SShapeElement implements Selectable, Alignable, Fadeable {
    text: string
    selected: boolean = false
    alignment: Point = ORIGIN_POINT
    opacity = 1

    hasFeature(feature: symbol) {
        return feature === boundsFeature || feature === alignFeature || feature === fadeFeature || feature === layoutableChildFeature
    }
}

/**
 * Serializable schema for SCompartment.
 */
export interface SCompartmentSchema extends SShapeElementSchema {
    layout?: string
}

/**
 * A compartment is used to group multiple child elements such as labels of a node. Usually a `vbox`
 * or `hbox` layout is used to arrange these children.
 */
export class SCompartment extends SShapeElement implements Fadeable {
    children: SChildElement[]
    layout?: string
    layoutOptions?: {[key: string]: string | number | boolean}
    opacity = 1

    hasFeature(feature: symbol) {
        return feature === boundsFeature || feature === layoutContainerFeature ||Â feature === layoutableChildFeature || feature === fadeFeature
    }
}

/**
 * A specialized model index that tracks outgoing and incoming edges.
 */
export class SGraphIndex extends SModelIndex<SModelElement> {

    private outgoing: Map<string, SEdge[]> = new Map
    private incoming: Map<string, SEdge[]> = new Map

    add(element: SModelElement): void {
        super.add(element)
        if (element instanceof SEdge) {
            // Register the edge in the outgoing map
            if (element.sourceId) {
                const sourceArr = this.outgoing.get(element.sourceId)
                if (sourceArr === undefined)
                    this.outgoing.set(element.sourceId, [element])
                else
                    sourceArr.push(element)
            }
            // Register the edge in the incoming map
            if (element.targetId) {
                const targetArr = this.incoming.get(element.targetId)
                if (targetArr === undefined)
                    this.incoming.set(element.targetId, [element])
                else
                    targetArr.push(element)
            }
        }
    }

    remove(element: SModelElement): void {
        super.remove(element)
        if (element instanceof SEdge) {
            // Remove the edge from the outgoing map
            const sourceArr = this.outgoing.get(element.sourceId)
            if (sourceArr !== undefined) {
                const index = sourceArr.indexOf(element)
                if (index >= 0) {
                    if (sourceArr.length === 1)
                        this.outgoing.delete(element.sourceId)
                    else
                        sourceArr.splice(index, 1)
                }
            }
            // Remove the edge from the incoming map
            const targetArr = this.incoming.get(element.targetId)
            if (targetArr !== undefined) {
                const index = targetArr.indexOf(element)
                if (index >= 0) {
                    if (targetArr.length === 1)
                        this.incoming.delete(element.targetId)
                    else
                        targetArr.splice(index, 1)
                }
            }
        }
    }

    getAttachedElements(element: SModelElement): SModelElement[] {
        const result: SModelElement[] = []
        const outgoing = this.outgoing.get(element.id)
        if (outgoing !== undefined) {
            outgoing.forEach(e => result.push(e))
        }
        const incoming = this.incoming.get(element.id)
        if (incoming !== undefined) {
            incoming.forEach(e => {
                if (outgoing === undefined || outgoing.indexOf(e) < 0) {
                    result.push(e)
                }
            })
        }
        return result
    }

    getIncomingEdges(node: SNode): Iterable<SEdge> {
        return this.incoming.get(node.id) || []
    }

    getOutgoingEdges(node: SNode): Iterable<SEdge> {
        return this.outgoing.get(node.id) || []
    }

}
