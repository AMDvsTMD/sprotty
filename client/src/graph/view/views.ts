import {h} from "snabbdom"
import {VNode} from "snabbdom/vnode"
import {View, RenderingContext} from "../../base/view"
import {Point} from "../../utils"
import {GGraph, GNode, GEdge} from "../model"

/**
 * View component that turns a GGraph element and its children into a tree of virtual DOM.
 */
export class GGraphView implements View {

    render(model: GGraph, context: RenderingContext): VNode {
        const vNode = h('svg', {
                key: model.id,
                class: {
                    graph: true
                },
                attrs: {
                    id: model.id
                }
            }, [
                h('g', {}, context.viewer.renderChildren(model, context))]
        );
        return vNode
    }
}

export abstract class GNodeView implements View {
    abstract render(model: GNode, context: RenderingContext): VNode

    abstract getAnchor(node: GNode, refPoint: Point, arrowLength: number)
}

export class StraightEdgeView implements View {
    render(edge: GEdge, context: RenderingContext) {
        const source = edge.source
        const target = edge.target
        if (source && target) {
            const sourceView = (context.viewer.viewRegistry.get(source.type, source)) as GNodeView
            const sourceAnchor = sourceView.getAnchor(source, target, 0)
            const targetView = (context.viewer.viewRegistry.get(target.type, target)) as GNodeView
            const targetAnchor = targetView.getAnchor(target, source, 0)
            const path = `M ${sourceAnchor.x},${sourceAnchor.y} L ${targetAnchor.x},${targetAnchor.y}`
            return h('path', {
                key: edge.id,
                class: {
                    edge: true
                },
                attrs: {
                    d: path,
                    id: edge.id,
                }
            })
        } else {
            return h('text', {
                key: edge.id,
                class: {
                    'dangling-edge': true
                },
                attrs: {
                    id: edge.id,
                }
            }, '?')
        }
    }
}

