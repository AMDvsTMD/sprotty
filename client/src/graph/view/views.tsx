import {View, RenderingContext} from "../../base/view"
import {Point, manhattanDistance} from "../../utils"
import {SGraph, SNode, SEdge} from "../model"
import {VNode} from "snabbdom/vnode"
import * as snabbdom from "snabbdom-jsx"

const JSX = {createElement: snabbdom.svg}

/**
 * View component that turns a SGraph element and its children into a tree of virtual DOM.
 */
export class SGraphView implements View {

    render(model: SGraph, context: RenderingContext): VNode {
        const transform = `scale(${model.zoom}) translate(${-model.scroll.x},${-model.scroll.y})`
        return <svg key={model.id} id={model.id} class-graph={true}>
            <g transform={transform}>
                {context.viewer.renderChildren(model, context)}
            </g>
        </svg>
    }
}

export abstract class SNodeView implements View {
    abstract render(model: SNode, context: RenderingContext): VNode

    abstract getAnchor(node: SNode, refPoint: Point): Point

    getWidth(node: SNode): number {
        return node.width;
    }

    getHeight(node: SNode): number {
        return node.height;
    }
}

export class StraightEdgeView implements View {
    render(edge: SEdge, context: RenderingContext) {
        const source = edge.source
        if (!source)
            return this.renderDanglingEdge("cannot resolve source", edge, context)

        const target = edge.target
        if (!target)
            return this.renderDanglingEdge("cannot resolve target", edge, context)

        const sourceView = context.viewer.viewRegistry.get(source.type, source)
        if (!(sourceView instanceof SNodeView))
            return this.renderDanglingEdge("expected source view type: SNodeView", edge, context)

        const targetView = context.viewer.viewRegistry.get(target.type, target)
        if (!(targetView instanceof SNodeView))
            return this.renderDanglingEdge("expected target view type: SNodeView", edge, context)

        let path: string
        if (edge.routingPoints && edge.routingPoints.length >= 1) {
            // Use the first routing point as start anchor reference
            let p0 = edge.routingPoints[0]
            const sourceAnchor = sourceView.getAnchor(source, p0)
            if (manhattanDistance(sourceAnchor, p0) > 2)
                path = `M ${sourceAnchor.x},${sourceAnchor.y} L ${p0.x},${p0.y}`
            else
                path = `M ${sourceAnchor.x},${sourceAnchor.y}`

            // Add the remaining routing points except the last one
            for (let i = 1; i < edge.routingPoints.length - 1; i++) {
                let pi = edge.routingPoints[i]
                path += ` L ${pi.x},${pi.y}`
            }
        } else {
            // Use the target center as start anchor reference
            const reference = {
                x: target.x + (targetView.getWidth(target) / 2 || 0),
                y: target.y + (targetView.getHeight(target) / 2 || 0)
            }
            const sourceAnchor = sourceView.getAnchor(source, reference)
            path = `M ${sourceAnchor.x},${sourceAnchor.y}`
        }

        if (edge.routingPoints && edge.routingPoints.length >= 2) {
            // Use the last routing point as end anchor reference
            let pn = edge.routingPoints[edge.routingPoints.length - 1]
            const targetAnchor = targetView.getAnchor(target, pn)
            if (manhattanDistance(targetAnchor, pn) > 2)
                path += ` L ${pn.x},${pn.y} L ${targetAnchor.x},${targetAnchor.y}`
            else
                path += ` L ${targetAnchor.x},${targetAnchor.y}`
        } else {
            // Use the source center as end anchor reference
            const reference = {
                x: source.x + (sourceView.getWidth(source) / 2 || 0),
                y: source.y + (sourceView.getHeight(source) / 2 || 0)
            }
            const targetAnchor = targetView.getAnchor(target, reference)
            path += ` L ${targetAnchor.x},${targetAnchor.y}`
        }

        return <path key={edge.id} id={edge.id} class-edge={true} d={path}/>
    }

    protected renderDanglingEdge(message: string, edge: SEdge, context: RenderingContext) {
        return <text key={edge.id} id={edge.id} class-dangling-edge={true} title={message}>?</text>
    }
}

