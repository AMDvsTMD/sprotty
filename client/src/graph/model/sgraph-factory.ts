import { injectable } from "inversify"
import { SModelFactory } from "../../base/model/smodel-factory"
import {
    getBasicType, SChildElement, SModelElementSchema, SModelRoot, SModelRootSchema, SParentElement
} from "../../base/model/smodel"
import { SEdge, SEdgeSchema, SGraph, SGraphSchema, SNode, SNodeSchema } from "./sgraph"

@injectable()
export class SGraphFactory extends SModelFactory {

    createElement(schema: SModelElementSchema, parent?: SParentElement): SChildElement {
        if (schema instanceof SChildElement) {
            if (parent !== undefined)
                schema.parent = parent
            return schema
        } else if (this.isNodeSchema(schema))
            return this.initializeChild(new SNode(), schema, parent)
        else if (this.isEdgeSchema(schema))
            return this.initializeChild(new SEdge(), schema, parent)
        else
            return super.createElement(schema, parent)
    }

    createRoot(schema: SModelRootSchema): SModelRoot {
        if (schema instanceof SModelRoot)
            return schema
        else if (this.isGraphSchema(schema))
            return this.initializeRoot(new SGraph(), schema)
        else
            return super.createRoot(schema)
    }

    isGraphSchema(schema: SModelElementSchema): schema is SGraphSchema {
        return getBasicType(schema) == 'graph'
    }

    isNodeSchema(schema: SModelElementSchema): schema is SNodeSchema {
        return getBasicType(schema) == 'node'
    }

    isEdgeSchema(schema: SModelElementSchema): schema is SEdgeSchema {
        return getBasicType(schema) == 'edge'
    }

}
