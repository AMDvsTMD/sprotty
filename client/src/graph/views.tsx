/*
 * Copyright (C) 2017 TypeFox and others.
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
 */

import * as snabbdom from 'snabbdom-jsx'
import { VNode } from "snabbdom/vnode"
import { Point, centerOfLine, maxDistance } from '../utils/geometry'
import { setAttr } from '../base/views/vnode-utils'
import { RenderingContext, IView } from "../base/views/view"
import { SModelElement, SParentElement } from "../base/model/smodel"
import { getSubType, translatePoint } from "../base/model/smodel-utils"
import { SRoutingHandle, isRoutable } from '../features/edit/model'
import { SCompartment, SEdge, SGraph, SLabel, SNode, SPort } from "./sgraph"
import { IAnchorableView, LinearRouter, RoutedPoint } from './routing'

const JSX = {createElement: snabbdom.svg}

/**
 * IView component that turns an SGraph element and its children into a tree of virtual DOM elements.
 */
export class SGraphView implements IView {

    render(model: SGraph, context: RenderingContext): VNode {
        const transform = `scale(${model.zoom}) translate(${-model.scroll.x},${-model.scroll.y})`
        return <svg class-sprotty-graph={true}>
            <g transform={transform}>
                {context.renderChildren(model)}
            </g>
        </svg>
    }
}

export abstract class AnchorableView implements IView, IAnchorableView {
    abstract render(model: SModelElement, context: RenderingContext): VNode

    abstract getAnchor(model: SNode | SPort, refPoint: Point, anchorCorrection: number): Point

    getStrokeWidth(model: SNode | SPort): number {
        return 0
    }

    getTranslatedAnchor(node: SNode | SPort, refPoint: Point, refContainer: SParentElement,
            anchorCorrection: number = 0, edge: SEdge): Point {
        const viewContainer = node.parent
        const anchor = this.getAnchor(node, translatePoint(refPoint, refContainer, viewContainer), anchorCorrection)
        const edgeContainer = edge.parent
        return translatePoint(anchor, viewContainer, edgeContainer)
    }
}

export class PolylineEdgeView implements IView {
    router = new LinearRouter() // TODO get via dependency injection

    render(edge: SEdge, context: RenderingContext): VNode {
        const source = edge.source
        if (source === undefined)
            return this.renderDanglingEdge("Cannot resolve source", edge, context)

        const target = edge.target
        if (target === undefined)
            return this.renderDanglingEdge("Cannot resolve target", edge, context)

        const sourceView = context.viewRegistry.get(source.type, source)
        if (!(sourceView instanceof AnchorableView))
            return this.renderDanglingEdge("Expected source view type: AnchorableView", edge, context)

        const targetView = context.viewRegistry.get(target.type, target)
        if (!(targetView instanceof AnchorableView))
            return this.renderDanglingEdge("Expected target view type: AnchorableView", edge, context)

        const route = this.router.route(edge, source, sourceView, target, targetView)

        return <g class-sprotty-edge={true} class-mouseover={edge.hoverFeedback}>
            {this.renderLine(edge, route, context)}
            {this.renderAdditionals(edge, route, context)}
            {context.renderChildren(edge, { route })}
        </g>
    }

    protected renderLine(edge: SEdge, segments: Point[], context: RenderingContext): VNode {
        const firstPoint = segments[0]
        let path = `M ${firstPoint.x},${firstPoint.y}`
        for (let i = 1; i < segments.length; i++) {
            const p = segments[i]
            path += ` L ${p.x},${p.y}`
        }
        return <path d={path}/>
    }

    protected renderAdditionals(edge: SEdge, segments: Point[], context: RenderingContext): VNode[] {
        return []
    }

    protected renderDanglingEdge(message: string, edge: SEdge, context: RenderingContext): VNode {
        return <text class-sprotty-edge-dangling={true} title={message}>?</text>
    }
}

export class SRoutingHandleView implements IView {
    minimalPointDistance: number = 10

    render(handle: SRoutingHandle, context: RenderingContext, args?: { route?: RoutedPoint[] }): VNode {
        if (args && args.route) {
            const position = this.getPosition(handle, args.route)
            handle.viewPosition = position
            if (position !== undefined) {
                const node = <circle class-sprotty-routing-handle={true}
                        class-selected={handle.selected} class-mouseover={handle.hoverFeedback}
                        cx={position.x} cy={position.y}/>   // Radius must be specified via CSS
                setAttr(node, 'data-kind', handle.kind)
                return node
            }
        }
        // Fallback: Create an empty group
        handle.viewPosition = undefined
        return <g/>
    }

    protected getPosition(handle: SRoutingHandle, route: RoutedPoint[]): Point | undefined {
        if (handle.kind === 'line') {
            const parent = handle.parent
            if (isRoutable(parent)) {
                const getIndex = (rp: RoutedPoint) => {
                    if (rp.pointIndex !== undefined)
                        return rp.pointIndex
                    else if (rp.kind === 'target')
                        return parent.routingPoints.length
                    else
                        return -1
                }
                let rp1, rp2: RoutedPoint | undefined
                for (const rp of route) {
                    const i = getIndex(rp)
                    if (i <= handle.pointIndex && (rp1 === undefined || i > getIndex(rp1)))
                        rp1 = rp
                    if (i > handle.pointIndex && (rp2 === undefined || i < getIndex(rp2)))
                        rp2 = rp
                }
                if (rp1 !== undefined && rp2 !== undefined) {
                    // Skip this handle if its related line segment is not included in the route
                    if (getIndex(rp1) !== handle.pointIndex && handle.pointIndex >= 0) {
                        const point = parent.routingPoints[handle.pointIndex]
                        if (maxDistance(point, rp1) >= maxDistance(point, rp2))
                            return undefined
                    }
                    if (getIndex(rp2) !== handle.pointIndex + 1 && handle.pointIndex + 1 < parent.routingPoints.length) {
                        const point = parent.routingPoints[handle.pointIndex + 1]
                        if (maxDistance(point, rp1) < maxDistance(point, rp2))
                            return undefined
                    }
                    // Skip this handle if its related line segment is too short
                    if (maxDistance(rp1, rp2) >= this.minimalPointDistance)
                        return centerOfLine(rp1, rp2)
                }
            }
        } else {
            return route.find(rp => rp.pointIndex === handle.pointIndex)
        }
        return undefined
    }
}

export class SLabelView implements IView {
    render(label: SLabel, context: RenderingContext): VNode {
        const vnode = <text class-sprotty-label={true}>{label.text}</text>
        const subType = getSubType(label)
        if (subType)
            setAttr(vnode, 'class', subType)
        return vnode
    }
}

export class SCompartmentView implements IView {
    render(model: SCompartment, context: RenderingContext): VNode {
        const translate = `translate(${model.bounds.x}, ${model.bounds.y})`
        const vnode = <g transform={translate} class-sprotty-comp="{true}">
            {context.renderChildren(model)}
        </g>
        const subType = getSubType(model)
        if (subType)
            setAttr(vnode, 'class', subType)
        return vnode
    }
}
