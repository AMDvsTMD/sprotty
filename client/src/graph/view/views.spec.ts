import {SGraphView, StraightEdgeView} from "./views"
import {SGraph, SEdge, SNode} from "../../graph/model/sgraph"
import {SGraphFactory} from "../model/sgraph-factory"
import {Viewer} from "../../base/view/viewer"
import {ViewRegistry} from "../../base/view/views"
import {CircleNodeView} from "../../../examples/circlegraph/src/views"
const toHTML = require('snabbdom-to-html')

describe('graph views', () => {

    const viewer = new Viewer()
    const viewRegistry = new ViewRegistry()
    viewer.viewRegistry = viewRegistry
    viewRegistry.register('graph', SGraphView)
    viewRegistry.register('node:circle', CircleNodeView)
    viewRegistry.register('edge:straight', StraightEdgeView)
    const context = {viewer: viewer}
    const factory = new SGraphFactory

    it('empty graph', () => {
        const schema = {
            type: 'graph',
            id: 'mygraph',
            children: []
        }
        const graph = factory.createRoot(schema) as SGraph
        const view = new SGraphView()
        const vnode = view.render(graph, context)
        const html = toHTML(vnode)
        expect(html).toMatchSnapshot()
    })

    const node0 = {id: 'node0', type: 'node:circle', x: 100, y: 100};
    const node1 = {id: 'node1', type: 'node:circle', x: 200, y: 150, selected: true};
    const edge0 = {id: 'edge0', type: 'edge:straight', sourceId: 'node0', targetId: 'node1'};
    const graph = factory.createRoot({id: 'graph', type: 'graph', children: [node0, node1, edge0]}) as SGraph;

    it('straight edge', () => {
        const view = new StraightEdgeView()
        const vnode = view.render(graph.index.getById('edge0') as SEdge, context)
        const html = toHTML(vnode)
        expect(html).toMatchSnapshot()
    })

    it('circle node', () => {
        const view = new CircleNodeView()
        const vnode = view.render(graph.index.getById('node0') as SNode, context)
        const html = toHTML(vnode)
        expect(html).toMatchSnapshot()
    })

    it('graph', () => {
        const view = new SGraphView()
        const vnode = view.render(graph, context)
        const html = toHTML(vnode)
        expect(html).toMatchSnapshot()
    })
})
