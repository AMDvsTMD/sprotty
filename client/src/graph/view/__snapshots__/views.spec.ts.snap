// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`graph views circle node 1`] = `"<circle id=\\"node0\\" class=\\"node\\" key=\\"node0\\" r=\\"40\\"></circle>"`;

exports[`graph views empty graph 1`] = `"<svg id=\\"mygraph\\" class=\\"graph\\"><g></g></svg>"`;

exports[`graph views graph 1`] = `"<svg id=\\"graph\\" class=\\"graph\\"><g><g transform=\\"translate(100, 100)\\"><circle id=\\"node0\\" class=\\"node\\" key=\\"node0\\" r=\\"40\\" style=\\"opacity: 1; transition: opacity 0.5s\\" /></g><g transform=\\"translate(200, 150)\\"><circle id=\\"node1\\" class=\\"node selected\\" key=\\"node1\\" r=\\"40\\" style=\\"opacity: 1; transition: opacity 0.5s\\" /></g><path id=\\"edge0\\" class=\\"edge\\" d=\\"M 135.77708763999664,117.88854381999832 L 164.22291236000336,132.11145618000168\\" style=\\"opacity: 1; transition: opacity 0.5s\\" /></g></svg>"`;

exports[`graph views straight edge 1`] = `"<path id=\\"edge0\\" class=\\"edge\\" d=\\"M 135.77708763999664,117.88854381999832 L 164.22291236000336,132.11145618000168\\"></path>"`;
